### PR TITLE
expression: provide `BuildExprWithAst` to build expression without planner context

### LINKED ISSUE #50463: introduce `BuildExprWithAst` to build expression without planner context
see descriptions in #50464

### PR DESCRIPTION
**Pull request by lcwangchao**
### What problem does this PR solve?

Issue Number: close #50463

### What changed and how does it work?

We provide a new function `expression.BuildExprWithAst` to build some simple expressions. Different with the exists `rewriteAstExpr`, it does not construct a plan builder when building expression. So it requires less context information than `rewriteAstExpr`.

`expression.BuildExprWithAst` is only used to build some simple expressions. When a expression contains a subquery, `BuildExprWithAst` will return an error.

In the later PRs, we should introduce `expression.BuildContext` to replace `sessionctx.Context` as the input context of `BuildExprWithAst`. So that we can build an expression outside a session without using mock context.

### Check List

Tests 

- [x] Unit test
- [ ] Integration test
- [ ] Manual test (add detailed scripts or steps below)
- [ ] No need to test
  > - [ ] I checked and no code files have been changed.
  > 

Side effects

- [ ] Performance regression: Consumes more CPU
- [ ] Performance regression: Consumes more Memory
- [ ] Breaking backward compatibility

Documentation

- [ ] Affects user behaviors
- [ ] Contains syntax changes
- [ ] Contains variable changes
- [ ] Contains experimental features
- [ ] Changes MySQL compatibility

### Release note

Please refer to [Release Notes Language Style Guide](https://pingcap.github.io/tidb-dev-guide/contribute-to-tidb/release-notes-style-guide.html) to write a quality release note.

```release-note
None
```

### CODE REVIEW

**Review by YangKeao — APPROVED** [review]
LGTM

File: pkg/planner/core/expression_rewriter.go [hunk 1]
```diff
@@ -85,6 +85,49 @@ func rewriteAstExpr(sctx sessionctx.Context, expr ast.ExprNode, schema *expressi
...
+
+	rewriter := &expressionRewriter{
+		ctx:                 context.TODO(),
+		sctx:                ctx,
+		schema:              options.InputSchema,
+		names:               options.InputNames,
+		sourceTable:         options.SourceTable,
+		allowBuildCastArray: options.AllowCastArray,
+		asScalar:            true,
+	}
```
[the remark this answers is not in the document]
Comment (lcwangchao) [reply pkg/planner/core/expression_rewriter.go:110]: I think it is not very necessary, at least for now. `BuildExprWithAst` is used to replace `RewriteAstExpr` which is implemented to build a planner every time and does not actually use the pool. For most times `RewriteAstExpr` is not in the critical path and does not affect the performance too much, we should avoid premature optimization.

[the remark this answers is not in the document]
Comment (lcwangchao) [reply pkg/planner/core/expression_rewriter.go:110]: It is simple too support pool if we want in the future, we can add a new option `WithRewriterPool` to provide pool for `BuildExprWithAst` and it is optional

**Review by winoros — APPROVED** [review]

**Review by XuHuaiyu — APPROVED** [review]

**Pull request merged**